Highlight and simplify mismatched types

estebank opened the pull request:
Shorten mismatched types errors by replacing subtypes that are not
different with `_`, and highlighting only the subtypes that are
different.

Given a file

```rust
struct X<T1, T2> {
    x: T1,
    y: T2,
}

fn foo() -> X<X<String, String>, String> {
    X { x: X {x: "".to_string(), y: 2}, y: "".to_string()}
}

fn bar() -> Option<String> {
    "".to_string()
}
```

provide the following output

```rust
error[E0308]: mismatched types
  --> file.rs:6:5
   |
 6 |     X { x: X {x: "".to_string(), y: 2}, y: "".to_string()}
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `std::string::String`, found {integer}
   |
   = note: expected type `X<X<_, std::string::String>, _>`
                                 ^^^^^^^^^^^^^^^^^^^   // < highlighted
              found type `X<X<_, {integer}>, _>`
                                 ^^^^^^^^^             // < highlighted

error[E0308]: mismatched types
  --> file.rs:6:5
   |
10 |     "".to_string()
   |     ^^^^^^^^^^^^^^ expected struct `std::option::Option`, found `std::string::String`
   |
   = note: expected type `Option<std::string::String>`
                          ^^^^^^^                   ^  // < highlighted
              found type `std::string::String`
```

Fix #21025. Re: #40186. Follow up to #39906.

I'm looking to change how this output is accomplished so that it doesn't create list of strings to pass around, but rather add an elided `Ty` placeholder, and use the same string formatting for normal types. I'll be doing that soonish.

r? @nikomatsakis

estebank committed:
Highlight and simplify mismatched types

Shorten mismatched types errors by replacing subtypes that are not
different with `_`, and highlighting only the subtypes that are
different.

Given a file

```rust
struct X<T1, T2> {
    x: T1,
    y: T2,
}

fn foo() -> X<X<String, String>, String> {
    X { x: X {x: "".to_string(), y: 2}, y: "".to_string()}
}

fn bar() -> Option<String> {
    "".to_string()
}
```

provide the following output

```rust
error[E0308]: mismatched types
  --> file.rs:6:5
   |
 6 |     X { x: X {x: "".to_string(), y: 2}, y: "".to_string()}
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `std::string::String`, found {integer}
   |
   = note: expected type `X<X<_, std::string::String>, _>`
                                 ^^^^^^^^^^^^^^^^^^^   // < highlighted
              found type `X<X<_, {integer}>, _>`
                                 ^^^^^^^^^             // < highlighted

error[E0308]: mismatched types
  --> file.rs:6:5
   |
10 |     "".to_string()
   |     ^^^^^^^^^^^^^^ expected struct `std::option::Option`, found `std::string::String`
   |
   = note: expected type `Option<std::string::String>`
                          ^^^^^^^                   ^  // < highlighted
              found type `std::string::String`
```

estebank commented:
Visual output:

<img width="1018" alt="" src="https://cloud.githubusercontent.com/assets/1606434/24884674/6376159c-1dff-11e7-952d-f35554949b89.png">

arielb1 commented:
@estebank 
Shouldn't the "found type `std::string::String`" be highlighted as well?

estebank commented:
It only highlights differing parts, while dimming the parts that are the same. In that case, only the `std::option::Option<...>` part is relevant to the error. May be we can incorporate the elision logic without any highlighting change for the "composition" case?

arielb1 commented:
@estebank 

I mean, I expected:
```

note: expected type `std::option::Option<std::string::String>`
                     ^^^^^^^^^^^^^^^^^^^^                   ^
      found type    `std::string::String`
                     ^^^^^^^^^^^^^^^^^^^
```

arielb1 commented on src/librustc/infer/error_reporting/mod.rs:
````diff
@@ -365,6 +365,257 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
+    /// The type error output will behave in the following way:
+    ///
+    /// ```text
+    /// Foo<Bar<Qux>
````
syntax error: mismatched parentheses

estebank replied:
Fixed.

arielb1 commented on src/librustc/infer/error_reporting/mod.rs:
````diff
@@ -365,6 +365,257 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
+    /// populate `other_value` with `other_ty`.
+    ///
+    /// ```text
+    /// Foo<Bar<Qux>
````
syntax error: mismatched parentheses

estebank replied:
Fixed.

arielb1 commented on src/librustc_errors/diagnostic.rs:
```diff
@@ -35,6 +35,46 @@ pub struct SubDiagnostic {
     pub render_span: Option<RenderSpan>,
 }
 
+#[derive(PartialEq, Eq)]
```
Call this `StyledString` or something?

estebank replied:
Done.

nikomatsakis commented on src/librustc/infer/error_reporting/mod.rs:
```diff
@@ -365,6 +365,257 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
+                        let l2 = lifetime_display(lifetimes.1);
+                        if l1 == l2 {
+                            values.0.push_normal("'_");
+                            values.1.push_normal("'_");
```
Unlike `_`, this is not legal syntax for a user to type, but I guess that's ok. Not sure what else to use, in any case.

estebank replied:
I agree with your reasoning. I followed this path given that the compiler already has diagnostics that show `'_` in some cases.

nikomatsakis commented on src/test/ui/mismatched_types/abridged.stderr:
```diff
@@ -0,0 +1,63 @@
+30 |     Foo { bar: 1 }
+   |     ^^^^^^^^^^^^^^ expected enum `std::option::Option`, found struct `Foo`
+   |
+   = note: expected type `std::option::Option<Foo>`
```
too bad we lose the highlighting here, but I don't really know how to fix that

estebank replied:
Yeah... I've been thinking that we could expand ui errors to have a mode where ANSI escape codes are kept (`.stderr_full`? A comment flag in the `.rs` test file?).

nikomatsakis reviewed (approved):
r=me modulo test failures; @arielb1's thoughts on underlining seem potentially fine, but I could go either way (and even if we did want to do it, I'd be fine w/ leaving it for a later PR).

nikomatsakis commented:
I see @arielb1 also posted some other suggestions that seem potentially good; e.g. using `StyledString` or what have you. I don't have a strong opinion. This didn't seem like a lot of code and it read fairly easily (thanks @estebank for the comments, the code seemed pretty easy to understand this time) but DRYer is always better.

arielb1 commented on src/librustc/infer/error_reporting/mod.rs:
```diff
@@ -365,6 +365,257 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
+                let mut values: (TypeString, TypeString) = (TypeString::new(), TypeString::new());
+                let path1 = self.tcx.item_path_str(def1.did.clone());
+                let path2 = self.tcx.item_path_str(def2.did.clone());
+                if path1 == path2 {
```
compare the dids?

estebank replied:
Done.

nikomatsakis replied:
yes, that seems better (@estebank -- just change to `if def1.did == def2.did {`)

estebank commented:
@nikomatsakis I believe I addressed all the points (except [the highlighting case brought up](https://github.com/rust-lang/rust/pull/41205#issuecomment-293247998) by @arielb1), as well as fixed a small presentational bug that was left from when I was first preparing this code.

estebank commented:
@bors r=nikomatsakis

bors commented:
:pushpin: Commit 2389830 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 2389830deaf39b736e35c471e39543d30f708972 -->
